### PR TITLE
Integration tests

### PR DESCRIPTION
**Pull request by funkyfuture**
This adds an initial framework for tests against larger sets of real-world corpora.

A mixed bag of fixes for discovered bugs comes with it.

Let me know if we should look at it together in person.

This PR has higher priority than #82. I still need to find the opportunity to create new graphics for that.

I'll continue to work on serialization based on the additions here.

### CODE REVIEW

**Comment by JKatzwinkel** [comment]
looks awesome! i'd need some time to set it up and check it out but i dont see why this shouldnt be merged. lets just set required coverage down to 96% to make the ci stop complaining!

**Review by JKatzwinkel — APPROVED** [review]

**Review by JKatzwinkel — APPROVED** [review]

**Comment by funkyfuture** [comment]
> looks awesome! i'd need some time to set it up and check it out but i dont see why this shouldnt be merged. lets just set required coverage down to 96% to make the ci stop complaining!

since it only affects Python 3.7:

- i'd have rather just disabled the coverage constraint for this version / or enable it only for the latest interpreter version
- my hope is still, that it would hit the expectation when the whole serialization code is there

it's okay for now though.

**Comment by funkyfuture** [comment]
> as an aside, i very much enjoy modeling a dsl for integration test scenarios (as you already did) and then extract scenario specs out of the test suite code and into yaml files early on. really adds to developing speed and overall perceived comfort in my experience. maybe keep something like this in mind for consideration sometime later?

i can't really follow. feel free to open an issue with a practice oriented sketch.

the next addition test-wise will be the comparison of a pretty serialisat in conjunction with collapsed whitespace.

File: integration-tests/fetch-corpora.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    url: str
+    archive_documents_root: str
+    target_directory: str
+    extra_files: dict[str, str] = {}
+
+
+CORPORA_PATH: Final = Path(__file__).parent.resolve() / "corpora"
+
+
+ARCHIVE_DESCRIPTIONS: Final = (
```
Comment (JKatzwinkel) [inline integration-tests/fetch-corpora.py:38]: i was talking about populating `ARCHIVE_DESCRIPTIONS` from a multi-document yaml file, but that's just an idea, there is no real need for that.

**Comment by JKatzwinkel** [comment]
> > looks awesome! i'd need some time to set it up and check it out but i dont see why this shouldnt be merged. lets just set required coverage down to 96% to make the ci stop complaining!
> 
> since it only affects Python 3.7:
> 
> * i'd have rather just disabled the coverage constraint for this version / or enable it only for the latest interpreter version
> * my hope is still, that it would hit the expectation when the whole serialization code is there
> 
> it's okay for now though.

personally, I can make my peace with a 96% coverage minimum, but don't let me stop you applying the gate selectively for all versions except 3.7 because I have no idea how one would do that.

File: integration-tests/fetch-corpora.py (hunk 1 above)
Comment (funkyfuture) [reply integration-tests/fetch-corpora.py:38]: i see. but the problem is that this model only covers one form of provided data.

e.g. i also investigated the "Hopper Texts", and not only would the download usually abort, the tarball consisted of further tarballs. i ditched it also due to the actual contents. the point is that diverse sources will be hard to describe in one model.

**Comment by funkyfuture** [comment]
> personally, I can make my peace with a 96% coverage minimum

well, it's a regression. but getting rid of the lxml backend will result in something very different code-volume-wise and 99% will be the target then.

and these tests increase edge case coverage a lot, that's way more important than a stupid number.

> because I have no idea how one would do that.

something with cli arguments i guess. but it's simple enough now.

File: integration-tests/fetch-corpora.py (hunk 1 above)
Comment (JKatzwinkel) [reply integration-tests/fetch-corpora.py:38]: i usually solve this by beriving the model from the file name or something similar, but right now the integration code is so nice to look at, we should leave it as it is anyway without making it needlessly complicated.

**Comment by funkyfuture** [comment]
thanks for the feedback. i hope the other branches rebased themselves when i arrive home.
